### PR TITLE
Return new metrics when creating metrics for a resource

### LINKED ISSUE #12: rest call for appending metrics to a resource does not return any result
See https://bugs.launchpad.net/gnocchi/+bug/1640942

### PR DESCRIPTION
**Pull request by BigChief45**
Fixes #12

### CODE REVIEW

File: gnocchi/rest/__init__.py [hunk 1]
```diff
@@ -665,6 +665,8 @@ def post(self):
         except indexer.NoSuchResource as e:
             abort(404, e)
 
+        return metrics
```
Comment (sileht) [inline gnocchi/rest/__init__.py]: We should not return the user input, but the values that come from the database. Also we update a resource, so we should return a resource description, not the metrics (the resource description should have the complete metrics info).

Maybe just adding 'return' at the beginning of line 655 is enough.

**Comment by BigChief45** [comment]
Are you referring to the [`test_post_append_metrics`](https://github.com/gnocchixyz/gnocchi/blob/master/gnocchi/tests/test_rest.py#L985) test?

**Comment by sileht** [comment]
On Wed, May 31, 2017 at 02:01:00AM -0700, Andrés Álvarez wrote:
>Are you referring to the [`test_post_append_metrics`](https://github.com/gnocchixyz/gnocchi/blob/master/gnocchi/tests/test_rest.py#L985) test?

Not only, you can looks at travis-ci job results to see all the failing
tests.

-- 
Mehdi Abaakouk
mail: sileht@sileht.net
irc: sileht

File: gnocchi/rest/__init__.py [hunk 2]
```diff
@@ -665,6 +665,9 @@ def post(self):
         except indexer.NoSuchResource as e:
             abort(404, e)
 
+        return pecan.request.indexer.get_resource(
```
Comment (jd) [inline gnocchi/rest/__init__.py]: If the user does not need this output it's a whole request for nothing.
It might  be better to leave the user request it if it needs it after, no?

File: gnocchi/tests/test_rest.py [hunk 3]
```diff
@@ -989,7 +989,7 @@ def test_post_append_metrics(self):
         metrics = {'foo': {'archive_policy_name': "high"}}
         self.app.post_json("/v1/resource/" + self.resource_type
                            + "/" + self.attributes['id'] + "/metric",
-                           params=metrics, status=204)
+                           params=metrics, status=200)
```
Comment (jd) [inline gnocchi/tests/test_rest.py]: This needs to be in the previous patch IIUC.

File: gnocchi/rest/__init__.py (hunk 2 above)
Comment (BigChief45) [inline gnocchi/rest/__init__.py]: Do you mean that the user should somehow specify if the output is needed? Could you show me an example of this?

Comment (jd) [inline gnocchi/rest/__init__.py]: Not really. What I mean is that if the user wants the new updated resource he just have to do `GET /v1/resource/generic/<id>`.

And actually returning the entire resource on the `/v1/resource/generic/<id>/metric` seems wrong to me. This is not the resource endpoint, it's the metric of the resource endpoint. So if anything would be returned it's the new list of metric only – which would be a different SQL request, not sure there's such a call in the indexer API.


Also I did understand at first read you were trying to fix #12. I'd like to have @chungg opionion since he is the one that opened the bug, if this is really a requirement since it will cost an indexer request on each call and I'm not sure it's really worth it.

Comment (BigChief45) [inline gnocchi/rest/__init__.py]: I agree. 

On my first commit before rebasing I was only returning the newly created metrics. But @sileht mentioned something about returning the whole resource's description, which would return the new metrics as well.

Waiting on more information to know how to proceed with this.

**Comment by sileht** [comment]
>On my first commit before rebasing I was only returning the newly created metrics. But @sileht mentioned something about returning the whole resource's description, which would return the new metrics as well.

Sorry for that, I have thought this API endpoint was about resource, but it's not.

-- 
Mehdi Abaakouk
mail: sileht@sileht.net
irc: sileht

**Comment by BigChief45** [comment]
Any consensus on what the response content should be? Or if this change should be abandoned.

**Comment by jd** [comment]
Re-reading this, I think your first patch was correct – @sileht misread the problem as he said after.

You should return the newly metrics that are created, so "return metrics" was the right choice.

But, as @sileht said you can't blindly return "metrics" because it's user-supplied definitions, and not metrics objects from the database.

`update_resource()` returns the resource with all its metric loaded. So I'd suggest actually returning `r.metrics` at the end of the function. That'll report all the metrics attached to this resource, which considering the endpoint URL, sounds like a good thing.

**Comment by BigChief45** [comment]
@jd Understood. I have made the changes. Will squash if everything is ok.

**Comment by sileht** [comment]
The solution looks good, maybe we should check the returned json with gabbi in additional to the 204 -> 200 change to make this PR perfect.

**Comment by BigChief45** [comment]
@sileht I have updated the Gabbi tests to check for the JSON response. Please let me know if these are correct and if they are enough.

**Comment by jd** [comment]
That's cool, though we need to squash all of that in a single commit. Can you do that @BigChief45? Otherwise I'll try that big UI button in GitHub. :)

**Comment by BigChief45** [comment]
@jd No problem. But I was thinking maybe you could try that big button this time? I've actually never seen it in action myself hehe.

**Comment by jd** [comment]
@BigChief45 Maybe next time, since you'll need to edit that branch to fix the Travis failures. ;)

**Comment by BigChief45** [comment]
@jd No problem. Fixing this soon.

By the way, I am taking a look at the documentation regarding this:

image

I cannot find the response content/scenario anywhere in the docs. Is this content generated into the docs from an actual API call?

**Comment by jd** [comment]
@BigChief45 Yes, the whole API doc is generated from real API calls. Check out https://github.com/gnocchixyz/gnocchi/blob/master/doc/source/rest.yaml and https://github.com/gnocchixyz/gnocchi/blob/master/doc/source/rest.j2

**Comment by BigChief45** [comment]
```
: Unable to match $[-1].resource_id as 85C44741-CC60-4033-804E-2D3098C7D2E9, got 85c44741-cc60-4033-804e-2d3098c7d2e9
```

Hm why are resource IDs being passed in uppercase in the Gabbi tests? Seems a bit strange.

**Comment by jd** [comment]
@BigChief45 Likely because whoever wrote that put them that way. It does not matter and the case should be preserved, so just upper case your changes and you should be all set.

File: gnocchi/rest/__init__.py [hunk 4]
```diff
@@ -665,6 +667,8 @@ def post(self):
         except indexer.NoSuchResource as e:
             abort(404, e)
 
+        return r.metrics
```
Comment (chungg) [inline gnocchi/rest/__init__.py]: i think this will return all metrics in resource not just the new metrics you added. it's failing in gate in gate because the new metric is not necessarily the first item returned.

if you're testing this locally, you can add verbose: true to gabbi tests, so check what order metrics are returned in (i'm guessing alphabetical). something like:

     - name: create new metrics
       verbose: True
       POST: /v1/resource/generic/f93450f2-d8a5-4d67-9985-02511241e7d1/metric
       request_headers:
           x-user-id: 0fbb231484614b1a80131fc22f6afc9c
           x-project-id: f3d41b770cc14f0bb94a1d5be9c0e3ea
           content-type: application/json
        data:
          foobar:
            archive_policy_name: low
       status: 200
       response_json_paths:
           $[0].name: foobar
           $[0].resource_id: f93450f2-d8a5-4d67-9985-02511241e7d1

Comment (BigChief45) [inline gnocchi/rest/__init__.py]: This issue is driving me nuts. Locally, all the gabbi tests pass, assuming that the metric being created in the test will be the first one to be returned in the list.

When testing using the Gnocchi API, metrics are not returned in alphabetical order, but in the order of creation (last created, first in the list). But I don't get why it is not the case in the gate.

**Comment by BigChief45** [comment]
So it seems that the order of the returned metrics differs between Postgresql and MySQL?

File: gnocchi/rest/__init__.py (hunk 4 above)
Comment (jd) [inline gnocchi/rest/__init__.py]: There's no order set by default. So, *by chance* it might always been the same order, but it also can depends on whetever it runs on MySQL or PostgreSQL. You can't assume in the test that the order is going to be always the same. Especially because maybe the next release of $RDBMS will break that assumption.

Now, what you can do is sort in the `response_json_paths`. This is documented here:

http://gabbi.readthedocs.io/en/latest/jsonpath.html

Basically you can do this:

```
response_json_paths:
  $[/name][1].name: foobar
```

If your metric list is "foobar, cpu_util" for example (you'll need to check what the full list is, then "foobar" will be alphabetically after "cpu_util" when sorting by name (the `[/name]` in the JSON path expression.

**Comment by BigChief45** [comment]
Seems like there is some issue with Docker in the CI.

**Comment by jd** [comment]
That should be fixed now.

**Comment by BigChief45** [comment]
I have squashed all the commits and all seems to pass at the gate. I think this is ready to be merged? :smiley:

**Comment by jd** [comment]
Thanks @BigChief45 !
